Fcl 199 async package party

dasormeter opened the pull request:
The WritePackagesService  and the WritePartyService  which must be implemented by each DAML-conforming ledger currently, for package management and party allocation, return a result type synchronously which is not a reliable indicator of successful processing or a descriptive indicator of failure.
https://github.com/digital-asset/daml/blob/9f9d2d0458013cb8f75dfc5c630a1b711c38359a/ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/WritePackagesService.scala#L44-L47

https://github.com/digital-asset/daml/blob/9f9d2d0458013cb8f75dfc5c630a1b711c38359a/ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/WritePartyService.scala#L37-L40

This PR moves these two interface to follow a similar pattern to transaction command submission, where SubmissionResult is returned for the write submission, and then a corresponding accept or reject event can be read asynchronously via the read service to confirm successful processing. This has a tremendous knock-on impact for persisting and processing of log entries.

The new SubmissionResult return type write interfaces as defined on this PR are :

https://github.com/digital-asset/daml/blob/FCL-199-async-package-party/ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/WritePackagesService.scala#L44-L47

https://github.com/digital-asset/daml/blob/FCL-199-async-package-party/ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/WritePartyService.scala#L37-L40

This PR introduces new  PackageUploadEntryAccept / PackageUploadEntryReject and PartyAllocationAccept and PartyAllocationReject events which should be used as confirmation of successful processing by the calling implementation.

These events must be persisted to DB for the indexer, and there should only be a single accept or reject event per PackageUpload request or PartyAllocation request.

The return type of the ledger api package management service and party management services are not changed as part of these PR, and the synchronous implementation of the ledger api service therefore needs to bring together the commit submission with the corresponding entry accept/reject to give full detail of error or party allocation when returning to the client using the ledger api services.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/unreleased.rst), if appropriate
- [x] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

dajmaki commented on ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/dao/JdbcLedgerDao.scala:
```diff
@@ -1300,6 +1300,33 @@ private class JdbcLedgerDao(
+      Future.failed,
+      _ =>
+        dbDispatcher.executeSql("store package rejections") { implicit conn =>
+          val sqlupdate = executeBatchSql(
```
this doesn't seem to be updating the external offset. If the indexer would die after processing this it would end up re-indexing this entry, which would likely fail because of a (participant_id, submission_id) uniqueness constraint, which would then end up making the indexer stuck (we can never proceed if we fail to index an entry as that might mean we're skipping over events, which would just end up failures down the line)

gerolf-da replied:
See https://github.com/digital-asset/daml/blob/9f9d2d0458013cb8f75dfc5c630a1b711c38359a/ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/dao/JdbcLedgerDao.scala#L708

dajmaki commented on ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/dao/JdbcLedgerDao.scala:
```diff
@@ -1385,6 +1413,12 @@ object JdbcLedgerDao {
         |on conflict (package_id) do nothing""".stripMargin
 
+    override protected[JdbcLedgerDao] val SQL_INSERT_PACKAGE_REJECTS: String =
+      """insert into package_rejects(package_id, participant_id, submission_id, reason)
```
I would follow the approach that we're doing with transactions: we have one table for per-submission kind that records both accepts and rejects.

dasormeter replied:
went with separate table

dajmaki commented on ledger/sandbox/src/main/resources/db/migration/postgres/V12__package_entries.sql:
```diff
@@ -0,0 +1,22 @@
+
+ALTER TABLE packages
+    -- SubmissionId for package to be uploaded
+    ADD submission_id varchar not null,
```
I don't think this approach works. You need a separate table for the package uploads as one package upload can upload multiple packages. I would add a table called "package_upload_entries" that records what happened to a submission.

dasormeter replied:
@dajmaki The thought was not wanting to duplicate the binary payload in yet another table as we will now always store whatever we have received.

But yes if we want to treat status of the multiple packages as one atomic entry the separate table would be cleaner

And would we still keep a duplicate in the packages table in its existing form to be used by listPackages and related api calls?

dajmaki replied:
There's no need to duplicate they payload. One table contains the actual packages and the other contains the submission results. You don't need the payload in the submission results table, you can just list the package ids that were actually uploaded by the submission.

dasormeter replied:
now created a separate package upload entries table (and party allocation entries table) that will store both accept and reject

dasormeter commented:
`s sandbox-scala-tests_test_suite_src_test_suite_scala_com_digitalasset_platform_sandbox_ScenarioLoadingITPostgres.scala ` looks to be blinky -- it passes locally but seems to be right on the limit of the 60s timeout given to it.  Rerunning the lot and also may be worth increasing the length given to that IT https://daholdings.slack.com/archives/CG1NG467K/p1574272893484700

dasormeter commented:
/azp run

stefanobaghino-da commented:
I deleted the description that points to a source which is not available to the wider public. Can you add the relevant bits from your internal ticket? If there are no sensitive information a copy-paste will do. :wink:

mziolekda commented on ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/services/admin/ApiPackageManagementService.scala:
```diff
@@ -74,30 +77,45 @@ class ApiPackageManagementService(
+            case SubmissionResult.NotSupported =>
+              Future.failed(ErrorFactories.unimplemented("not supported"))
+            case SubmissionResult.Overloaded =>
+              Future.failed(ErrorFactories.grpcError(io.grpc.Status.RESOURCE_EXHAUSTED))
```
Please use
`Future.failed(ErrorFactories.resourceExhausted(r.description))` instead

dasormeter replied:
Changed to use this -- had to add description to SubmissionResult case objects as they did not have this field previously

mziolekda commented on ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/ReadOnlySqlLedger.scala:
```diff
@@ -122,7 +122,7 @@ private class ReadOnlySqlLedgerFactory(
       case Some(initialId) =>
         ledgerDao
-          .lookupLedgerId()
+          .lookupLedgerId("_initialize_some")
```
Do you intend to keep those?

dasormeter replied:
these are gone now

dasormeter commented on ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/InMemoryKVParticipantState.scala:
```diff
@@ -176,8 +173,8 @@ class InMemoryKVParticipantState(
 
-      case AddPackageUploadRequest(submissionId, cf) =>
-        packageRequests += (submissionId -> cf); ()
+//      case AddPackageUploadRequest(submissionId, cf) =>
```
addressed in latest push

dasormeter commented on ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/InMemoryKVParticipantState.scala:
```diff
@@ -209,13 +206,7 @@ class InMemoryKVParticipantState(
                   partyRequests -= submissionId
 
                 case KeyValueConsumption.PackageUploadResponse(submissionId, result) =>
```
addressed in latest push

gerolf-da commented on ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/Update.scala:
```diff
@@ -59,12 +59,16 @@ object Update {
       participantId: ParticipantId,
-      recordTime: Timestamp)
+      recordTime: Timestamp,
+      submissionId: String)
```
We have different wrapped types for the IDs instead of just using `String` directly. See `ApplicationId`, `CommandId`, `PackageId`, etc...

dasormeter replied:
@gerolf-da do we have any type for submissionId?  this is kinda like a CommandId but isn't there a typical format convention for CommandId, or is it just an opaque string?  typing submissionId as commandId may be a bit misleading though

gerolf-da replied:
You can introduce a `SubmissionId` the same way as `CommandId`. And yes, they are just opaque strings (adhering to the `LedgerString` restrictions).

gerolf-da replied:
Use `SubmissionId`?

dasormeter replied:
addressed

gerolf-da commented on ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/services/admin/ApiPackageManagementService.scala:
```diff
@@ -74,30 +77,45 @@ class ApiPackageManagementService(
     )
   }
 
+  private def prepareUploadPackageResult(
```
This is not called anywhere.

dasormeter replied:
deleted this for now -- will add back any relevant similar parsing when polling is implemented

gerolf-da commented on ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/dao/JdbcLedgerDao.scala:
```diff
@@ -81,9 +81,9 @@ private class JdbcLedgerDao(
+  override def lookupLedgerId(callerId: String): Future[Option[LedgerId]] =
     dbDispatcher
-      .executeSql("get_ledger_id") { implicit conn =>
+      .executeSql(s"get_ledger_id$callerId") { implicit conn =>
```
What purpose does the `callerId` serve?

dasormeter replied:
was using this for some initialization instrumentation : now removed

gerolf-da commented on ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/services/admin/ApiPackageManagementService.scala:
```diff
@@ -74,30 +77,45 @@ class ApiPackageManagementService(
           .toScala(res._1)
           .flatMap {
-            case UploadPackagesResult.Ok =>
+            //TODO BH: first step to get this to compile -- need to implement properly polling and retrieval of response event
```
TODO: Implement polling based on package upload entries and not the list of packages stored.

stefanobaghino-da replied:
Who's `BH`? :thinking:

dasormeter replied:
yeah this one is still pending -- will tackle in the a.m.

dasormeter replied:
implemented now

dasormeter commented on ledger/sandbox/src/main/resources/db/migration/h2database/V8__Party_allocation_entries.sql:
```diff
@@ -0,0 +1,25 @@
+CREATE TABLE party_allocation_entries
+(
+    -- SubmissionId for the party allocation
+    submission_id    varchar not null,
```
now added

dasormeter commented:
> I deleted the description that points to a source which is not available to the wider public. Can you add the relevant bits from your internal ticket? If there are no sensitive information a copy-paste will do. 😉

dasormeter commented on ledger/sandbox/src/main/resources/db/migration/postgres/V13__Party_allocation_entries.sql:
```diff
@@ -0,0 +1,25 @@
+CREATE TABLE party_allocation_entries
+(
+    -- SubmissionId for the party allocation
+    submission_id    varchar not null,
```
added

dasormeter commented:
whoops - closed by mistake by hitting the wrong button :(

dasormeter commented on ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/InMemoryKVParticipantStateIT.scala:
```diff
@@ -225,17 +227,9 @@ class InMemoryKVParticipantStateIT
+            fail("unexpected response to party allocation.  Error : " + result.description)
         }
       }
-//      ps.allocateParty(hint, displayName).thenApply[Assertion]({
```
i don't see this commented out line : maybe outddated?

dasormeter replied:
@mziolekda i went through the history on this commented code and cannot see it ever running or being applicable for the "accept allocateParty when hint is empty" test so don't think we want to enable it.  Do you think there is a missing testcase though?

mziolekda replied:
you are right, drop it

dasormeter commented on ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueConsumption.scala:
```diff
@@ -44,31 +44,53 @@ object KeyValueConsumption {
     entry.getPayloadCase match {
       case DamlLogEntry.PayloadCase.PACKAGE_UPLOAD_ENTRY =>
+        val participantId =
+          Ref.LedgerString.assertFromString(entry.getPackageUploadEntry.getParticipantId)
```
have made changes to only send one accept message (if there are multiple packages) and if all the packages are duplicate, to still send the accept.  IT updated for InMemoryParticipantIT

dasormeter replied:
> Don't generate that update when the participantId does not match though.
I see the participant on the log entry but don't see an easy way of discovering the running participant in KeyValueConsumption.scala : are you saying to include this check in the JdbcIndexer before persisting it?

mziolekda replied:
Leave that for the moment, leave a TODO instead, we have the same problem across the board right now

dasormeter replied:
okay i looked at this again and see it should just be a case of passing through participantId instance variable from the specific participant state that is calling logEntryToUpdate, but will leave it as a TODO for now

dasormeter commented on ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueConsumption.scala:
```diff
@@ -44,31 +44,53 @@ object KeyValueConsumption {
+          Ref.LedgerString.assertFromString(entry.getPackageUploadEntry.getParticipantId)
+        val rejection = entry.getPackageUploadRejectionEntry
+        val proposedPackageUpload = rejection.getInvalidPackage
+        entry.getPackageUploadEntry.getArchivesList.asScala.map { archive =>
```
updated to only send one rejected update per entry (rather than per package)

miklos-da commented on ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueConsumption.scala:
```diff
@@ -44,31 +41,56 @@ object KeyValueConsumption {
+              case DamlPackageUploadRejectionEntry.ReasonCase.REASON_NOT_SET =>
+                s"Package ${proposedPackageUpload.getDetails} upload failed for undetermined reason"
+              case DamlPackageUploadRejectionEntry.ReasonCase.PARTICIPANT_NOT_AUTHORIZED =>
+                s"Participant ${participantId} not authorized to upload package"
```
You don't need `{}` for interpolating a variable reference.

dasormeter replied:
resolved

miklos-da commented on ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KeyValueConsumption.scala:
```diff
@@ -123,23 +145,14 @@ object KeyValueConsumption {
-      participantId: String): Option[AsyncResponse] = {
+      participantId: ParticipantId): Option[AsyncResponse] = {
 
     entry.getPayloadCase match {
```
To improve readability, this may now be rewritten with `PartialFunction.condOpt` if we don't explicitly want to have a `case` for each `PayloadCase` (as most of the responses will be `None`):
```scala
PartialFunction.condOpt(entry.getPayloadCase) {
  case DamlLogEntry.PayloadCase.PARTY_ALLOCATION_ENTRY if participantId == entry.getPartyAllocationEntry.getParticipantId =>
    PartyAllocationResponse(
              entry.getPartyAllocationEntry.getSubmissionId,
              ...
    )
  case DamlLogEntry.PayloadCase.PARTY_ALLOCATION_REJECTION_ENTRY if participantId == entry.getPartyAllocationEntry.getParticipantId =>
    partyRejectionEntryToAsyncResponse(entry.getPartyAllocationRejectionEntry)
  ...
}
```

dasormeter replied:
Thanks - will try this in future.  In this case this function will soon go away once the party allocation entries are moved to the new style so will leave it for the moment and remove once the party allocation entries are gone

dasormeter replied:
this method is now deleted

miklos-da commented on ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/SubmissionResult.scala:
```diff
@@ -3,20 +3,32 @@
-  final case object Acknowledged extends SubmissionResult
+  final case object Acknowledged extends SubmissionResult {
+    override def description: String =
+      s"The request has been received"
```
Unnecessary string interpolation.

dasormeter replied:
resolved

miklos-da commented on ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/dao/JdbcLedgerDao.scala:
```diff
@@ -83,7 +84,7 @@ private class JdbcLedgerDao(
   override def lookupLedgerId(): Future[Option[LedgerId]] =
     dbDispatcher
-      .executeSql("get_ledger_id") { implicit conn =>
+      .executeSql(s"get_ledger_id") { implicit conn =>
```
We don't need string interpolation here, right?

miklos-da commented on ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/dao/JdbcLedgerDao.scala:
```diff
@@ -1307,6 +1346,59 @@ private class JdbcLedgerDao(
+      participantId: ParticipantId,
+      submissionId: String,
+      reason: Option[String],
+      typ: String): Future[PersistenceResponse] = {
```
Please make the accept/reject case typed and use e.g. `case object`s or an `Enumeration` instead of `typ: String`.

gerolf-da replied:
The `typ` can also be derived from whether a `reason` is `Some` or `None`, no?

dasormeter replied:
added use of PackageUploadEntry type

miklos-da commented on ledger/sandbox/src/main/resources/db/migration/h2database/V7__Package_upload_entries.sql:
```diff
@@ -0,0 +1,35 @@
+    participant_id   varchar            not null,
+    -- The type of entry, one of 'accept' or 'reject'
+    typ              varchar            not null,
+    -- If the type is 'rejection', then the rejection reason is set.
```
`'reject'`

dasormeter replied:
resolved

miklos-da commented on ledger/sandbox/src/main/scala/com/digitalasset/platform/index/JdbcIndexer.scala:
```diff
@@ -273,6 +273,30 @@ class JdbcIndexer private[index] (
+            participantId,
+            submissionId,
+            None,
+            "accept")
```
To avoid possible inconsistencies, please switch to a typed accept/reject representation (or, alternatively, define constants for such magic strings).

dasormeter replied:
addressed now, and put the magic strings as value of the corresponding case classes

gerolf-da commented on ledger/participant-state/kvutils/tools/src/main/scala/com/daml/ledger/participant/state/kvutils/tools/IntegrityCheck.scala:
```diff
@@ -115,7 +115,10 @@ object IntegrityCheck extends App {
+        assert(
+          updates.forall(update =>
+            update.isInstanceOf[Update.PublicPackageUploaded] || update
+              .isInstanceOf[Update.PackageUploadEntryAccepted]))
```
Does this also need the clause for `Update.PackageUPloadEntryRejected`?

dasormeter replied:
yes good point -- that should be there too but probably was not exercised / evident because the semantic test that is run to gather this ledger dump does not include any PackageUploadEntryRejected items

dasormeter replied:
added Rejected and typ is now typed

gerolf-da commented on ledger/sandbox/src/main/scala/com/digitalasset/platform/index/JdbcIndexer.scala:
```diff
@@ -350,12 +374,14 @@ class JdbcIndexer private[index] (
       case ConfigurationChanged(_, _) => None
       case ConfigurationChangeRejected(_, _) => None
       case CommandRejected(_, _) => None
+      case PackageUploadEntryAccepted(_, _) => None
```
Why does the accept message not contain a record time, similar to `TransactionAccepted`?

dasormeter replied:
added recordTime to these Update events too

dajmaki replied:
heads-up that my config indexing PR adds the recordTime to the abstract base class.

dasormeter commented:
/azp run

gerolf-da commented on ledger/ledger-api-common/src/main/scala/com/digitalasset/ledger/package.scala:
```diff
@@ -22,4 +22,7 @@ package object ledger {
   type EventId = EventId.T
 
+  /** Identifiers used to correlate admin submission with results. */
+  val SubmissionId: LedgerString.type = LedgerString
```
This should only be defined in the `participant-state` module.

dasormeter replied:
removed SubmissionId from these two modules 
![image](https://user-images.githubusercontent.com/25179017/69777864-12582f80-1170-11ea-9f79-d3f1816713e0.png)

gerolf-da commented on ledger/sandbox/src/main/resources/db/migration/h2database/V7__Package_upload_entries.sql:
```diff
@@ -0,0 +1,35 @@
+-- SPDX-License-Identifier: Apache-2.0
+
+---------------------------------------------------------------------------------------------------
+-- V12: package_upload_entries
```
This is `V7` for H2.

dasormeter replied:
corrected

gerolf-da commented on ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/dao/JdbcLedgerDao.scala:
```diff
@@ -1307,6 +1346,68 @@ private class JdbcLedgerDao(
+      endExclusive,
+      PageSize,
+      (startI, endE) => {
+        dbDispatcher.executeSql(s"load package upload entries", Some(s"bounds: [$startI, $endE[")) {
```
```suggestion
        dbDispatcher.executeSql(s"load_package_upload_entries", Some(s"bounds: [$startI, $endE[")) {
```

dasormeter replied:
addressed

gerolf-da commented on ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/dao/JdbcLedgerDao.scala:
```diff
@@ -1307,6 +1346,68 @@ private class JdbcLedgerDao(
+      externalOffset: Option[ExternalOffset],
+      entry: PackageUploadEntry): Future[PersistenceResponse] = {
+
+    dbDispatcher.executeSql("store package upload entry") { implicit conn =>
```
```suggestion
    dbDispatcher.executeSql("store_package_upload_entry") { implicit conn =>
```

dasormeter replied:
addressed

gerolf-da commented on ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/dao/JdbcLedgerDao.scala:
```diff
@@ -1307,6 +1346,68 @@ private class JdbcLedgerDao(
+    }
+  }
+
+  private def reasonOrNull(entry: PackageUploadEntry): String = {
```
Doesn't seem to be used anywhere.

dasormeter replied:
it is used again now

gerolf-da commented:
This PR is in conflict with #3553 regarding the database migration versions.

dasormeter commented:
/azp run

dasormeter commented:
/azp help

gerolf-da commented on ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/SqlLedger.scala:
```diff
@@ -297,22 +302,64 @@ private class SqlLedger(
+      participantId: ParticipantId): Future[SubmissionResult] = {
+
+    val storePartyAllocationEntry: Future[SubmissionResult] = {
+      var headRef = 0L
```
Party allocation entries need to go through the same "enqueuing" mechanism as the transaction ledger entries.

dasormeter replied:
enqueue now there on the branch.  NOTE: i had to reduce the number of parties in the LotsOfParties test to 125 as the previoius 1024 were failing in the two LotsOfParties tests throwing a mess of RESOURCE_EXHAUSTED grpc overload errors.

spoke with @stefanobaghino-da this morning and said we will temporarily reduce to get this merged in, but will need to investigate why this behaves differently when using enqueue, and pump back up to higher number once we have found the issue and resolved in a separate PR

gerolf-da commented on ledger/participant-state-index/src/main/scala/com/daml/ledger/participant/state/index/v2/IndexPartyManagementService.scala:
```diff
@@ -16,4 +18,10 @@ trait IndexPartyManagementService {
 
   def listParties(): Future[List[PartyDetails]]
+
+  def getPartyAllocationEntries(
```
This doesn't seem to be used anywhere.

dasormeter replied:
gone

gerolf-da commented on ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/PackageUploadLedgerEntry.scala:
```diff
@@ -5,27 +5,25 @@ package com.digitalasset.platform.sandbox.stores.ledger
 }
 
-object PackageUploadEntry {
+object PackageUploadLedgerEntry {
```
The entries are missing record time.

dasormeter replied:
now includes recordTime and wired in to store this to DB

gerolf-da commented on ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/PartyAllocationLedgerEntry.scala:
```diff
@@ -0,0 +1,33 @@
+import com.digitalasset.ledger.api.domain.PartyDetails
+
+//TODO BH consider removing duplication of this class
+sealed abstract class PartyAllocationLedgerEntry() extends Product with Serializable {
```
The entries are missing record time fields.

dasormeter replied:
recordTime added

gerolf-da commented on ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/dao/JdbcLedgerDao.scala:
```diff
@@ -1377,49 +1434,123 @@ private class JdbcLedgerDao(
+      PageSize,
+      (startI, endE) => {
+        dbDispatcher
+          .executeSql(s"load party allocation entries", Some(s"bounds: [$startI, $endE[")) {
```
Please don't use spaces in the first parameter to `executeSql`.

dasormeter replied:
resolved

gerolf-da commented on ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/dao/JdbcLedgerDao.scala:
```diff
@@ -1377,49 +1434,123 @@ private class JdbcLedgerDao(
+        SQL(queries.SQL_INSERT_PARTY_ALLOCATION_ENTRY)
+          .on(
+            "ledger_offset" -> offset,
+            "recorded_at" -> Instant.now(),
```
This needs to use `recordTime` from the allocation entries.

dasormeter replied:
now corrected

dasormeter commented on ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/inmemory/InMemoryLedger.scala:
```diff
@@ -75,6 +75,12 @@ class InMemoryLedger(
   }
 
+  private val allocationEntries = {
+    val l = new LedgerEntries[PartyAllocationLedgerEntry](_.toString)
```
hmm this will need some generalizing as the LedgerEntry type has a bunch of stuff that is not relevant for the PackageUpload and PartyAllocation entries.  Let's chat on this tomorrow

dasormeter replied:
this is now addressed through @dajmaki change to have `InMemoryPartyEntry` and `InMemoryPackageEntry` extend `InMemoryEntry` a la InMemoryLedgerEntry and use the same ledgerEntries ImmArray to store items, preserving ordering and offset

dasormeter commented on ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/services/admin/ApiPartyManagementService.scala:
```diff
@@ -52,58 +58,50 @@ class ApiPartyManagementService private (
+      submissionId: SubmissionId): Future[domain.PartyAllocationEntry] = {
+    PollingUtils
+      .pollSingleUntilPersisted(() =>
+        partyManagementService.lookupPartyAllocationEntry(submissionId))(
```
that stream was the initial idea to be able to return all but it ended up being more than was needed since, at the moment, really we only need single lookup by submissionId so i will probably kill getPartyAllocationEntries (preferably as a follow-on PR)

dasormeter replied:
i've killed getPartyAllocationEntries and partyAllocationEntries on this PR now

dasormeter commented:
> This PR is in conflict with #3553 regarding the database migration versions.

Reversioned to V8/V9 and V13/V14 to avoid clash with V7 and V12

dasormeter commented on ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/services/admin/ApiPackageManagementService.scala:
```diff
@@ -59,13 +60,14 @@ class ApiPackageManagementService(
   }
 
   override def uploadDarFile(request: UploadDarFileRequest): Future[UploadDarFileResponse] = {
+    val submissionId = UUID.randomUUID().toString
```
also used further down in the polling call -- putting only in yield block it won't be in scope for the later call
![image](https://user-images.githubusercontent.com/25179017/69778618-face7600-1172-11ea-8d5b-4f99cf6bf385.png)

dasormeter commented on ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/services/admin/ApiPartyManagementService.scala:
```diff
@@ -25,7 +28,10 @@ class ApiPartyManagementService private (
     scheduler: Scheduler,
-    loggerFactory: NamedLoggerFactory
+    loggerFactory: NamedLoggerFactory,
+    implicit val ec: ExecutionContext,
```
removed these

dasormeter commented on ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/SandboxIndexAndWriteService.scala:
```diff
@@ -413,26 +437,36 @@ class LedgerBackedWriteService(ledger: Ledger, timeProvider: TimeProvider) exten
+          ledger.allocateParty(party, displayName, submissionId, participantId))
       case Some(Left(error)) =>
-        CompletableFuture.completedFuture(PartyAllocationResult.InvalidName(error))
+        // TODO BH : don't think we want to throw submission result error here but rather submit the request
```
okay i will leave as is for now then and remove the TODO

dasormeter commented:
> This PR is in conflict with #3553 regarding the database migration versions.

now merged with master by @dajmaki  incorporating all these conflicting changes

gerolf-da commented on ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/InMemoryPackageStore.scala:
```diff
@@ -50,6 +52,15 @@ case class InMemoryPackageStore(
+  override def lookupPackageUploadEntry(
+      submissionId: SubmissionId): Future[Option[domain.PackageUploadEntry]] = {
+    // TODO BH : not clear what codepath gets you here, but
+    // extending IndexPackagesService requires this method to be implemented
```
There's actually no reason for this class to extend `IndexPackagesService`. Please remove that.

dasormeter replied:
woohoo!  that's what i was hoping for

dasormeter replied:
marcin has already removed -- marking as resolved

gerolf-da commented on ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/MeteredLedger.scala:
```diff
@@ -75,10 +77,18 @@ private class MeteredReadOnlyLedger(ledger: ReadOnlyLedger, metrics: MetricRegis
 
+  override def lookupPackageUploadEntry(
+      submissionId: SubmissionId): Future[Option[PackageUploadLedgerEntry]] =
+    ledger.lookupPackageUploadEntry(submissionId)
```
Add a metric for this method.

gerolf-da commented on ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/PackageUploadLedgerEntry.scala:
```diff
@@ -0,0 +1,34 @@
+  val submissionId: String
+  val participantId: ParticipantId
+  val recordTime: Instant
+  def value: String
```
`value` is not a particularly descriptive name. Please follow how it was done for `ConfigurationEntry`.

gerolf-da commented on ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/PartyAllocationLedgerEntry.scala:
```diff
@@ -0,0 +1,38 @@
+  val submissionId: String
+  val participantId: ParticipantId
+  val recordTime: Instant
+  def value: String
```
Same comment as for `PackageUploadLedgerEntry`. Please follow the `ConfigurationEntry` pattern.

gerolf-da commented on ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/SqlLedger.scala:
```diff
@@ -314,42 +320,105 @@ private final class SqlLedger(
+          None,
+          submissionId,
+          participantId,
+          //TODO BH proper participant isLocal check needed
```
What do you mean with this TODO?

gerolf-da commented on ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/SqlLedger.scala:
```diff
@@ -314,42 +320,105 @@ private final class SqlLedger(
-      }(DEC)
+
+    def storePackageUploadEntry(offsets: SqlLedger.Offsets) = {
+      ledgerDao
```
Now the storing of the package and the storing of the packageUploadEntry happen in a different database transaction, but this should rather happen atomically. See `JdbcLedgerDao#storeConfigurationEntry`.

dasormeter replied:
these are separate Update message though -- if there is a DAR with 3 packages there will be 3 PublicPackageUploaded messages (one for each package) + a 4th message for the accept of the lot.  In the case of any of the packages being rejected there will be 0 PublicPackageUploaded and 1 PackageUploadRejected

submissionId is currently the common identifier across these messages but you don't know when you receive the first PublicPackageUploaded how many more are to come and when the Accept/Reject will come.

When each of the entries are processed in kv land they will call logEntryToUpdate in KeyValueConsumption and fire the updates

We could potentially wrap all the PublicPackageUploaded in the PackageUploadAccepted message and process them as a batch to get atomicity

dasormeter replied:
@gerolf-da can i request that we track this as a follow-on PR as it will have some significant knock-on effect to bundle these together and guarantee atomicity

gerolf-da commented on ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/dao/JdbcLedgerDao.scala:
```diff
@@ -1464,6 +1559,136 @@ private class JdbcLedgerDao(
+      case PartyAllocationLedgerEntry.Rejected(_, _, _, _) =>
+        None
+    }
+  private def partyAllocationReasonorNull(entry: PartyAllocationLedgerEntry): String = {
```
```suggestion
  private def partyAllocationReasonOrNull(entry: PartyAllocationLedgerEntry): String = {
```

gerolf-da commented on ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/dao/LedgerDao.scala:
```diff
@@ -137,6 +147,17 @@ trait LedgerReadDao extends AutoCloseable with ReportsHealth {
+    * @param endExclusive   ending offset exclusive
+    * @return a stream of ledger entries tupled with their offset
+    */
+  def getPackageUploadEntries(
```
Not used anywhere. Please remove.

gerolf-da commented on ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/dao/LedgerDao.scala:
```diff
@@ -242,9 +271,42 @@ trait LedgerWriteDao extends AutoCloseable with ReportsHealth {
+    * @param entry
+    * @return
+    */
+  def storePackageUploadEntry(
```
Should be combined with `uploadLfPackages`.

dasormeter replied:
right now uploadLfPackages is called once per archive (in the case of a submission with multiple DARs) and then storePackageUploadEntry is called once at the end.  Or in the case of rejected packages uploadLfPackages is never called and storePackageUploadEntry is called once for the reject.

So similar to the above about atomicity can we track a change here as a follow-on PR?

dasormeter replied:
i take it back this one is doable - we basically should do the package upload and the insert of the entries within the one `uploadLfPackages` method

dasormeter replied:
i did most of the changes for this locally but what we don't have in teh PublicPackageUploaded event is the info on whether the package is accepted or not
![image](https://user-images.githubusercontent.com/25179017/70110356-b59bbf80-161c-11ea-8397-b64035719267.png)

gerolf-da commented on ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/dao/MeteredLedgerDao.scala:
```diff
@@ -82,9 +92,25 @@ private class MeteredLedgerReadDao(ledgerDao: LedgerReadDao, metrics: MetricRegi
+
+  override def lookupPackageUploadEntry(
+      submissionId: SubmissionId): Future[Option[PackageUploadLedgerEntry]] =
+    ledgerDao.lookupPackageUploadEntry(submissionId)
```
Missing metric.

gerolf-da commented on ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/dao/MeteredLedgerDao.scala:
```diff
@@ -184,6 +210,28 @@ private class MeteredLedgerDao(ledgerDao: LedgerDao, metrics: MetricRegistry)
+      newLedgerEnd: LedgerOffset,
+      externalOffset: Option[ExternalOffset],
+      entry: PackageUploadLedgerEntry): Future[PersistenceResponse] =
+    ledgerDao.storePackageUploadEntry(offset, newLedgerEnd, externalOffset, entry)
```
Missing metric.

gerolf-da commented on ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/dao/MeteredLedgerDao.scala:
```diff
@@ -184,6 +210,28 @@ private class MeteredLedgerDao(ledgerDao: LedgerDao, metrics: MetricRegistry)
+      submissionId: SubmissionId,
+      participantId: ParticipantId,
+      entry: PartyAllocationLedgerEntry): Future[PersistenceResponse] =
+    ledgerDao.storePartyAllocationEntry(
```
Missing metric.

dajmaki commented on ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/Update.scala:
```diff
@@ -107,17 +144,85 @@ object Update {
+    * Consider whether an enumerated set of reject reasons a la [[RejectionReason]] would be helpful, and whether the same breadth of reject
+    * types needs to be handled for package upload entry rejects
+    */
+  final case class PackageUploadEntryAccepted(
```
One is about command completion of an upload of a set of packages and the other is an event about upload of a single package. Submitting participant sees both, others just see the latter. We could consider refactoring this to the same form as the TransactionAccepted update with optional information about the submission.

dajmaki commented on ledger/sandbox/src/main/scala/com/digitalasset/platform/index/JdbcIndexer.scala:
```diff
@@ -273,6 +307,25 @@ class JdbcIndexer private[index] (
         )
         ledgerDao.uploadLfPackages(uploadId, packages, externalOffset).map(_ => ())(DEC)
 
+      case PackageUploadEntryAccepted(participantId, recordTime, submissionId) =>
```
One package upload submission can carry many packages and hence can emit multiple events, whereas party allocation only emits one event.

mziolekda replied:
Due to deduplication a low level package upload massage may result in 0..n packages. Now we have a message that represents the operation status (PackageUploadAccepted, whose target is the originating participant) and individual packages (PublicPackageUploaded)

dasormeter replied:
explanation given : marking as resolved for tracking as no change required (just explanation)

miklos-da commented on ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/Update.scala:
```diff
@@ -71,17 +71,54 @@ object Update {
+    * with a possible need to revisit as part of the per-party package visibility work
+    * https://github.com/digital-asset/daml/issues/311.
+    *
+    * @param participantId
```
The order here seems to be out of line with the actual parameter ordering -- you might want to fix that.

miklos-da commented on ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/Update.scala:
```diff
@@ -107,17 +144,85 @@ object Update {
+      submissionId: SubmissionId)
+      extends Update {
+    override def description: String = {
+      s"Accepted package ${submissionId} for participant ${participantId}"
```
You don't need curly braces here in the interpolated string.

miklos-da commented on ledger/sandbox/src/main/resources/db/migration/h2database/V8__Party_allocation_entries.sql:
```diff
@@ -0,0 +1,38 @@
+    -- displayName
+    display_name     varchar,
+    -- The type of entry, one of 'accept' or 'reject'
+    typ              varchar             not null,
```
`entry_type`?

dasormeter replied:
`typ` is the naming that was used for ledger_entries and configuration_entries so stuck with that in case we want to collapse these to a single table in future, so they are as compatible in naming as possible

i agree entry_type is clearer but i think we should change all at the same time if we decide to, and we'll need to do a migration script for the other tables if we rename

miklos-da replied:
Agreed -- in that case shall we add a technical task for that into our backlog?

dasormeter replied:
> Agreed -- in that case shall we add a technical task for that into our backlog?

Yes let's do that

miklos-da commented on ledger/sandbox/src/main/resources/db/migration/h2database/V9__Package_upload_entries.sql:
```diff
@@ -0,0 +1,35 @@
+    -- participant id that initiated the package upload
+    participant_id   varchar            not null,
+    -- The type of entry, one of 'accept' or 'reject'
+    typ              varchar            not null,
```
`entry_type`?

dasormeter replied:
same comment as above -- staying close to column naming for ledger_entries and configuration_entries

miklos-da commented on ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/InMemoryPackageStore.scala:
```diff
@@ -50,6 +52,15 @@ case class InMemoryPackageStore(
+    // extending IndexPackagesService requires this method to be implemented
+    logger.error(
+      s"this will never return  as we are unable to match to accept/reject package upload entry for  $submissionId")
+    Future.successful((None))
```
Unnecessary `()`.

miklos-da commented on ledger/ledger-api-domain/src/main/scala/com/digitalasset/ledger/api/domain.scala:
```diff
@@ -290,4 +290,50 @@ object domain {
+  sealed abstract class PartyAllocationEntry() extends Product with Serializable {
+    val submissionId: String
+    val participantId: ParticipantId
+    def value: String
```
Isn't it required indirectly for storing a value for the `entry_type` column in the DB? If yes, this should be a constant, if not, I agree with @gerolf-da.

gerolf-da commented on ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/Update.scala:
```diff
@@ -71,17 +71,54 @@ object Update {
+    *   The submissionId of the command which requested party to be added.
+    *
     */
+  // TODO BH: add isLocal as was on async event
```
The `isLocal` attribute can be figured out by the indexer itself (by comparing its own local participantId with the one provided in this message).

dasormeter replied:
addressed in latest changes for isLocal

gerolf-da commented on ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/Update.scala:
```diff
@@ -107,17 +144,85 @@ object Update {
+    * @param submissionId
+    *   submissionId of the package upload command.
+    *
+    * Consider whether an enumerated set of reject reasons a la [[RejectionReason]] would be helpful, and whether the same breadth of reject
```
This comment doesn't belong to the `PackageUploadAccepted` update message.

dasormeter replied:
resolved

gerolf-da commented on ledger/sandbox/src/main/scala/com/digitalasset/platform/index/JdbcIndexer.scala:
```diff
@@ -258,12 +262,42 @@ class JdbcIndexer private[index] (
-      case PublicPackageUploaded(archive, sourceDescription, _, _) =>
+        //store party allocation accepted entry in party allocation entries table too
+        ledgerDao
+          .storePartyAllocationEntry(
```
`storeParty` and `storePartyAllocationEntry` should be a single call, similar to how storing a `ledgerEntry` for a DAML transaction is done in the same database transaction as updating contracts, etc...

dasormeter replied:
`storeParty` is now gone and the `parties` and `party_allocation_entries` tables have been collapsed into `party_entries` so there is now just a single transaction.

gerolf-da commented on ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v1/Update.scala:
```diff
@@ -107,17 +144,85 @@ object Update {
+    * Consider whether an enumerated set of reject reasons a la [[RejectionReason]] would be helpful, and whether the same breadth of reject
+    * types needs to be handled for package upload entry rejects
+    */
+  final case class PackageUploadAccepted(
```
@mziolekda: can we go through the reasoning again why `PublicPackageUploaded` and `PackageUploadAccepted` are separate messages?

dasormeter replied:
PublicPackageUploaded is per package (if there is a dar with multiple packages, there will be three PublicPackageUploaded messages)

PackageUploadAccepted is for the entire submission (so for all 3 packages in the example)

If packages have already been uploaded (it is a duplicate upload), there will be no PublicPackageUploaded message and there will be a single PackageUploadAccepted message

If there are multiple packages and any of them is invalid or rejected, there will be no PublicPackageUploaded messages and there will be a single PackageUploadRejected message

gerolf-da commented on ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/MeteredLedger.scala:
```diff
@@ -34,6 +34,9 @@ private class MeteredReadOnlyLedger(ledger: ReadOnlyLedger, metrics: MetricRegis
     val lookupTransaction = metrics.timer("Ledger.lookupTransaction")
     val lookupLedgerConfiguration = metrics.timer("Ledger.lookupLedgerConfiguration ")
     val parties = metrics.timer("Ledger.parties")
+    val partyAllocationEntries = metrics.timer("Ledger.partyAllocationEntries")
```
Not used anywhere anymore.

dasormeter replied:
resolved

gerolf-da commented on ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/dao/JdbcLedgerDao.scala:
```diff
@@ -1388,6 +1400,16 @@ private class JdbcLedgerDao(
           |where package_id = {package_id}
           |""".stripMargin)
 
+  private val SQL_SELECT_PACKAGE_UPLOAD_ENTRIES =
```
Not used anywhere.

dasormeter replied:
resolved

gerolf-da commented on ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/dao/JdbcLedgerDao.scala:
```diff
@@ -1388,6 +1400,16 @@ private class JdbcLedgerDao(
+      "select * from package_upload_entries where ledger_offset>={startInclusive} and ledger_offset<{endExclusive} order by ledger_offset asc")
+
+  private val SQL_SELECT_PACKAGE_UPLOAD_ENTRY =
+    SQL("select * from package_upload_entries where submission_id={submissionId} limit 1")
```
What's the reason for having `limit 1` in the query?

dasormeter replied:
just because there should only be one entry for a single submission and we only need one entry returned for confirmation of either accept or reject processing

gerolf-da commented on ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/dao/JdbcLedgerDao.scala:
```diff
@@ -1388,6 +1400,16 @@ private class JdbcLedgerDao(
+    SQL("select * from package_upload_entries where submission_id={submissionId} limit 1")
+
+  private val SQL_SELECT_PARTY_ALLOCATION_ENTRY =
+    SQL("select * from party_allocation_entries where submission_id={submissionId} limit 1")
```
What's the reason for having `limit 1` in the query?

dasormeter replied:
just because there should only be one entry for a single submission and we only need one entry returned for confirmation of either accept or reject processing

gerolf-da commented on ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/dao/JdbcLedgerDao.scala:
```diff
@@ -1464,6 +1559,126 @@ private class JdbcLedgerDao(
+          .execute()
+        PersistenceResponse.Ok
+      }).recover {
+        case NonFatal(e) if e.getMessage.contains(queries.DUPLICATE_KEY_ERROR) =>
```
I don't think the insert query results in a duplicate error being thrown (seeing that it contains `on conflict do nothing`).

dasormeter replied:
resolved - just show full error now and removed the guard -- should never get here but if we do it will give a more descriptive error

dasormeter commented:
/azp run digital-asset.daml

dasormeter commented:
/azp run

dasormeter commented:
latest updates and pushes:

- parties and party_allocation_entries tables collapsed to a single party_entries table and all tests passing correctly (root cause of the linux blink i believe was a hard-coded None externalOffset which was resolved by looking up the externalOffset rather than hardcoding)
- use enqueue for allocateParty -- had to reduce the number of parties in LotsOfParties.scala to 125 (from 1024) because anything above that is throwing continual RESOURCE_EXHAUSTED messages when enqueue is used. https://github.com/digital-asset/daml/pull/3391/commits/13fde112b5235583aa7ca41ab15e7281e8fad326 . I would like to merge this and work together with @stefano.baghino and/or @samir to investigate what may be going wrong as a follow-up PR to get this back up to 1024
- as part of collapsing the parties table had to deal with implicit parties added as divulged parties in sandbox scenarios.  added another entry typ for implicit and renamed accept and reject to AllocationAccepted and AllocationRejected, and had to make some fields nullable that will not be there for implicit parties (namely participantId, since an implicit party is not really allocated to a particular participant).  
- renamed party allocation entry.... to party entry in many case classes and methods since it now includes implicit parties as well as allocated parties.  the remaining things called partyAllocation should only be relevant for the true allocation scenarios.
- as part of killing the parties table i also cleaned up this V4_1_CollectParties.scala that was reading from parties sql table, however i don't see any tests or usages of it so cannot really check if all the changes are kosher, or whether this file is really needed.  changes are here: https://github.com/digital-asset/daml/pull/3391/commits/ad78e9fc5b6ae039d72f0d0f72fee6c7e9a16175
- isLocal has correct logic for checks to compare participantId sent on message against participantId of running component.  derived value is now stored to DB and read back correctly rather than defaulting to true in various places.
- remove duplicate submissionId and the not-always-required participantId on storePartyEntry since there is no different behaviour for implicit party vs the party allocation entry items
- various cleanup and removal of dead code and addressing specific points raised by gerolf

dasormeter commented:
This is now superceded by https://github.com/digital-asset/daml/pull/3744 (party mgmt) and https://github.com/digital-asset/daml/pull/3806 (package mgmt) which have both been merged to master
